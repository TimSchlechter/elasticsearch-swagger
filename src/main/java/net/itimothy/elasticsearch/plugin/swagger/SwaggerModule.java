package net.itimothy.elasticsearch.plugin.swagger;

import net.itimothy.elasticsearch.plugin.swagger.rest.CategoriesRequestHandler;
import net.itimothy.elasticsearch.routes.ModelsCatalog;
import net.itimothy.elasticsearch.restapispec.OfficialRestApiSpecDataProvider;
import org.elasticsearch.common.inject.AbstractModule;
import net.itimothy.elasticsearch.plugin.swagger.v1_2.rest.ApiDeclarationRequestHandler;
import net.itimothy.elasticsearch.plugin.swagger.v1_2.rest.ResourceListingRequestHandler;

public class SwaggerModule extends AbstractModule {

    @Override
    protected void configure() {
        bind(CategoriesRequestHandler.class).asEagerSingleton();

        bind(ResourceListingRequestHandler.class).asEagerSingleton();
        bind(ApiDeclarationRequestHandler.class).asEagerSingleton();

        bind(ModelsCatalog.class).asEagerSingleton();
        bind(OfficialRestApiSpecDataProvider.class).asEagerSingleton();
    }

}
