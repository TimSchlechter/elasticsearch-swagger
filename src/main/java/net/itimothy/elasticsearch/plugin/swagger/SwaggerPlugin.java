package net.itimothy.elasticsearch.plugin.swagger;

import org.elasticsearch.common.inject.Module;
import org.elasticsearch.plugins.Plugin;

import java.util.Collection;
import java.util.Collections;

public class SwaggerPlugin extends Plugin {
    public final static String API_DOCS_PATH = "_swagger";

    @Override
    public String name() {
        return "swagger";
    }

    @Override
    public String description() {
        return "Swagger for Elasticsearch";
    }

    @Override
    public Collection<Module> nodeModules() {
        return Collections.<Module>singletonList(new SwaggerModule());
    }

}
