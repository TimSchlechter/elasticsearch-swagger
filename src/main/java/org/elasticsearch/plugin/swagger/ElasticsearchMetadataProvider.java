package org.elasticsearch.plugin.swagger;

import net.itimothy.rest.metadata.model.HttpMethod;
import net.itimothy.rest.metadata.MetadataProvider;
import net.itimothy.rest.metadata.model.Parameter;
import net.itimothy.rest.metadata.model.ParameterType;
import net.itimothy.rest.metadata.model.Route;

import java.util.List;

import static java.util.Arrays.asList;

/**
 * Provides Elasticsearch's API metadata in a generic format
 */
public class ElasticsearchMetadataProvider implements MetadataProvider
{
    ModelManager models = new ModelManager();

    @Override
    public String getVersion() {
        return "1.4";
    }

    @Override
    public List<Route> getRoutes() {
        return asList(
            Route.builder()
                .resourcePath("Mapping")
                .method(HttpMethod.GET)
                .apiPath("_mapping")
                .name("getMapping")
                .model(ModelManager.MAPPING)
                .build(),

            Route.builder()
                .resourcePath("Mapping")
                .method(HttpMethod.PUT)
                .parameters(asList(
                    Parameter.builder()
                        .paramType(ParameterType.BODY)
                        .model(ModelManager.MAPPING)
                        .build()
                ))
                .apiPath("_mapping")
                .name("putMapping")
                .model(ModelManager.MAPPING)
                .build()
        );
    }
}
