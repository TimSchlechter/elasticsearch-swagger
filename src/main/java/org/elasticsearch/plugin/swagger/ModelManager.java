package org.elasticsearch.plugin.swagger;

import net.itimothy.rest.metadata.model.Model;

import static java.util.Arrays.asList;

public class ModelManager {
    public static final Model MAPPING = Model.builder()
        .id("mapping")
        .properties(asList(

        ))
        .build();

}
