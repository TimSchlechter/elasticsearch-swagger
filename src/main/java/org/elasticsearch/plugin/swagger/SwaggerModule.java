package org.elasticsearch.plugin.swagger;

import org.elasticsearch.common.inject.AbstractModule;
import org.elasticsearch.plugin.swagger.v1_2.rest.ApiDeclarationRequestHandler;
import org.elasticsearch.plugin.swagger.v1_2.rest.ResourceListingRequestHandler;

public class SwaggerModule extends AbstractModule {

    @Override
    protected void configure() {
        bind(ResourceListingRequestHandler.class).asEagerSingleton();
        bind(ApiDeclarationRequestHandler.class).asEagerSingleton();
    }

}
