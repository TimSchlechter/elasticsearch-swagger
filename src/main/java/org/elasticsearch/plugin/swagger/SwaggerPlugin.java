package org.elasticsearch.plugin.swagger;

import org.elasticsearch.common.inject.Module;
import org.elasticsearch.plugins.AbstractPlugin;

import java.util.Arrays;
import java.util.Collection;

public class SwaggerPlugin extends AbstractPlugin {
    public final static String API_DOCS_PATH = "_swagger";

    @Override
    public String name() {
        return "swagger";
    }

    @Override
    public String description() {
        return "Swagger for Elasticsearch";
    }

    @Override
    public Collection<Class<? extends Module>> modules() {
        return Arrays.asList(SwaggerModule.class);
    }

}
