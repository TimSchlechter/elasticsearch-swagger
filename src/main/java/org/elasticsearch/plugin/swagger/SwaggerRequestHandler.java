package org.elasticsearch.plugin.swagger;

import org.elasticsearch.client.Client;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.plugin.swagger.v1_2.model.SwaggerModel;
import org.elasticsearch.rest.*;

import static org.elasticsearch.rest.RestStatus.OK;

public abstract class SwaggerRequestHandler extends BaseRestHandler {

    protected SwaggerRequestHandler(Settings settings, RestController controller, Client client) {
        super(settings, controller, client);
    }

    protected abstract SwaggerModel handleRequest(RestRequest request, Client client) throws Exception;

    @Override
    protected void handleRequest(RestRequest request, RestChannel channel, Client client) throws Exception {
        SwaggerModel model = handleRequest(request, client);
        RestResponse response = new BytesRestResponse(OK, "application/json", model.toJson());

        response.addHeader("Access-Control-Allow-Origin", "*");
        response.addHeader("Access-Control-Allow-Methods", "GET");
        response.addHeader("Access-Control-Allow-Headers", "Content-Type, api_key, Authorization");

        channel.sendResponse(response);
    }
}
