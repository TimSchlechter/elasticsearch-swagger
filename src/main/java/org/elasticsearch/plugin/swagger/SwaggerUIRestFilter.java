package org.elasticsearch.plugin.swagger;

import org.elasticsearch.http.netty.NettyHttpRequest;
import org.elasticsearch.rest.RestChannel;
import org.elasticsearch.rest.RestFilter;
import org.elasticsearch.rest.RestFilterChain;
import org.elasticsearch.rest.RestRequest;

import java.lang.reflect.Field;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static org.elasticsearch.plugin.swagger.SwaggerPlugin.PLUGIN_PATH;

public class SwaggerUIRestFilter extends RestFilter {

    @Override
    public void process(RestRequest request, RestChannel channel, RestFilterChain filterChain) throws Exception {
        String path = request.path();
        if (path.startsWith("/" + PLUGIN_PATH) && !path.startsWith("/" + PLUGIN_PATH + "/api-docs")) {
            Pattern p = Pattern.compile("^/" + PLUGIN_PATH + "/(.*)?");
            Matcher m = p.matcher(request.path());
            if (m.matches()) {
                Field pathField = NettyHttpRequest.class.getDeclaredField("rawPath");
                pathField.setAccessible(true);
                pathField.set(request, "_swagger");
                request.params().put("file", m.group(1));
            }
        }
        filterChain.continueProcessing(request, channel);
    }
}