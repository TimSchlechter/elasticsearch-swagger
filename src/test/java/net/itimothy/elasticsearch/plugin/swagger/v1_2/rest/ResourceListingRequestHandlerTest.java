package net.itimothy.elasticsearch.plugin.swagger.v1_2.rest;

import net.itimothy.elasticsearch.plugin.swagger.SwaggerApprovalTest;
import org.approvaltests.reporters.JunitReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.Test;

@UseReporter(JunitReporter.class)
public class ResourceListingRequestHandlerTest extends SwaggerApprovalTest {
    /*
    @Test
    public void get_cleanInstance_approves() throws Exception {
        verify("v1.2");
    }

    @Test
    public void getIndex_indexCreated_approves() throws Exception {
        client.admin().indices().prepareCreate("some-index").execute().actionGet();
        verify("some-index/v1.2");
    }
    */
}
