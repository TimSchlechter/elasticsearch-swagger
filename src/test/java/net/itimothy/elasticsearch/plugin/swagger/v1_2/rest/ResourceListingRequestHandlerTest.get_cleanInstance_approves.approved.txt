{
  "swaggerVersion": "1.2",
  "apis": [
    {
      "path": "/Index management"
    },
    {
      "path": "/Mapping management"
    },
    {
      "path": "/Alias management"
    },
    {
      "path": "/Index settings"
    },
    {
      "path": "/Search"
    }
  ],
  "apiVersion": "1.4.4",
  "info": {
    "title": "Elasticsearch 1.4.4 API"
  }
}